cli/cmd/check.go

```diff
@@ -216,14 +224,106 @@ func runChecks(w io.Writer, hc *healthcheck.HealthChecker) bool {
 			}
 		}
 
-		fmt.Fprintf(w, "%s %s\n", status, result.Description)
+		fmt.Fprintf(wout, "%s %s\n", status, result.Description)
 		if result.Err != nil {
-			fmt.Fprintf(w, "    %s\n", result.Err)
+			fmt.Fprintf(wout, "    %s\n", result.Err)
 			if result.HintAnchor != "" {
-				fmt.Fprintf(w, "    see %s%s for hints\n", healthcheck.HintBaseURL, result.HintAnchor)
+				fmt.Fprintf(wout, "    see %s%s for hints\n", healthcheck.HintBaseURL, result.HintAnchor)
 			}
 		}
 	}
 
-	return hc.RunChecks(prettyPrintResults)
+	success := hc.RunChecks(prettyPrintResults)
+	// this empty line separates final results from the checks list in the output
+	fmt.Fprintln(wout, "")
+
+	if !success {
+		fmt.Fprintf(wout, "Status check results are %s\n", failStatus)
+	} else {
+		fmt.Fprintf(wout, "Status check results are %s\n", okStatus)
+	}
+
+	return success
+}
+
+type checkOutput struct {
+	Success    bool             `json:"success"`
+	Categories []*checkCategory `json:"categories"`
+}
+
+type checkCategory struct {
+	Name   string   `json:"categoryName"`
+	Checks []*check `json:"checks"`
+}
+
+// check is a user-facing version of `healthcheck.CheckResult`, for output via
+// `linkerd check -o json`.
+type check struct {
+	Description string      `json:"description"`
+	Hint        string      `json:"hint,omitempty"`
+	Error       string      `json:"error,omitempty"`
+	Result      checkResult `json:"result"`
+}
+
+type checkResult string
+
+const (
+	checkSuccess checkResult = "success"
+	checkWarn    checkResult = "warning"
+	checkErr     checkResult = "error"
+)
+
+func runChecksJSON(wout io.Writer, werr io.Writer, hc *healthcheck.HealthChecker) bool {
+	var categories []*checkCategory
+
+	collectJSONOutput := func(result *healthcheck.CheckResult) {
+		categoryName := string(result.Category)
+		if categories == nil || categories[len(categories)-1].Name != categoryName {
+			categories = append(categories, &checkCategory{
+				Name:   categoryName,
+				Checks: []*check{},
+			})
+		}
+
+		if !result.Retry {
+			currentCategory := categories[len(categories)-1]
+			// ignore checks that are going to be retried, we want only final results
+			status := checkSuccess
+			if result.Err != nil {
+				status = checkErr
+				if result.Warning {
+					status = checkWarn
+				}
+			}
+
+			currentCheck := &check{
+				Description: result.Description,
+				Result:      status,
+			}
+
+			if result.Err != nil {
+				currentCheck.Error = result.Err.Error()
+
+				if result.HintAnchor != "" {
+					currentCheck.Hint = fmt.Sprintf("%s%s", healthcheck.HintBaseURL, result.HintAnchor)
+				}
+			}
+			currentCategory.Checks = append(currentCategory.Checks, currentCheck)
+		}
+	}
+
+	result := hc.RunChecks(collectJSONOutput)
+
+	outputJSON := checkOutput{
+		Success:    result,
+		Categories: categories,
+	}
+
+	resultJSON, err := json.MarshalIndent(outputJSON, "", "  ")
+	if err == nil {
+		fmt.Fprint(wout, string(resultJSON))
```

Comment:
apologies i missed this on the last pass, mind adding a trailing newline to the output, for consistency with the other commands?
```go
fmt.Fprintf(wout, "%s\n", resultJSON)
```